File: 277-Find-Celebrity/findCelebrity.java
public class Solution extends Relation {
  private int num_people;

  public int findCelebrity(int n) {
    num_people = n;
    if (n <= 1) {
      return n - 1;
    }

    int candidate = 0;

    for (int i = 1; i < n; i++) {
      if (knows(candidate, i)) {
        candidate = i;
      }
    }

    if (isCelebrity(candidate)) {
      return candidate;
    }
    return -1;
  }

  private boolean isCelebrity(int candidate) {
    for (int i = 0; i < num_people; i++) {
      if ((knows(candidate, i) || !knows(i, candidate)) && candidate != i) {
        return false;
      }
    }
    return true;
  }
}

File: 277-Find-Celebrity/findCelebrity.py
class Solution:
  def findCelebrity(self, n: int) -> int:
    if n <= 1:
      return n - 1
    
    candidate = 0
    
    for i in range(1, n):
      if knows(candidate, i):
        candidate = i
        
    return candidate if self.isCelebrity(candidate, n) else -1
  
  def isCelebrity(self, candidate, n):
    for i in range(n):
      if (knows(candidate, i) or not knows(i, candidate)) and i != candidate:
        return False
    return True